### PR TITLE
Updated SOS2 Patch

### LINKED ISSUE #3173: [Patch Request]:SOS2 Ship Shield Do not Intercept CE Projectiles
**Read and complete the template below. Requests that fail to follow the template will be deleted without warning.**

**If a Patch Request (open or closed) already exists for the same mod, you may add a reaction to indicate your support. Do not create another request. Duplicate requests will be deleted.**

**Title should be:** [Patch Request]: (Mod Name)

**Mod name: Save Our Ship 2**

**Link to mod download (preferably Steam workshop or Github):https://steamcommunity.com/sharedfiles/filedetails/?id=1909914131**

**Description:**
Provide a brief overview of what content the mod adds or modifies. Specify what content in particular you believe requires a patch to be compatibly with Combat Extended.
Xml patch is under way, found a issue that all CE projectiles bypasses SOS2 shield, would need a C# fix.
**Content:**
What does the new mod contain? Check all that apply.
- [x] New weapons.
- [x] New animals, humanoids, or other creatures.
- [ ] New factions.
- [x] New armor or apparel.
I'm working on the XML patch, no other issue so far.

**Complete the following checklist before submitting a request.**
I hereby verify that I have done all of the following:
- [x] Searched the Steam workshop, Discord, and Github for an existing patch and not found one.
- [x] Confirmed there are no other open Patch Requests for this mod.
- [x] Confirmed the requested mod is not known to be fundamentally incompatible with Combat Extended.
- [x] **Have completed this template accurately and completely.**
- [x] **Understand that putting in a Patch Request does not guarantee a patch will be made.**
- [x] **Agree that I will not directly contact contributors to ask about patch requests.**

**Putting in request does not guarantee a patch will be made. Do not directly contact team members about Patch Requests. For smaller mods, consider reviewing the guide and making a patch yourself. If you wish to create a patch for any requested mod, there is no need to ask for permission.**

### PR DESCRIPTION
**Pull request by Rhinous**
## Additions

- Added support for SOS2 shields to intercept `ProjectileCE`
- Patched ship weapons GroundDefenseMode by creating a new building `Building_ShipTurretCE` and new verb `Verb_ShootShipCE` that uses CE logic on the ground while keeping the SOS2 logic for space combat
  - Only small ship turrets allow shooting on ground to match SOS2
  - Ship turrets shoot from above the roof height but their projectiles are not treated as mortars
  - Ship turret projectiles draw without a height offset to match the SOS2 turret graphics. The default CE height offset looked really weird with the flat turret graphics since the shot position is from above roof height
## Changes

- Updated SOS2 Reference assembly
- Updated Vehicle Framework Reference Assembly
- Updated SOS2 Torpedo tube reload patch
- Moved all SOS2 patching to its own project similar to the VehicleFramework patch
- Added `virtual` to `TryCollideWithRoof` in `ProjectileCE` so that the roof collision logic can be overwritten, stopping the player from accidentally blowing up their own spaceship's roof if the shot misses its target
- Changed the access level of a few fields/methods in `Verb_LaunchProjectileCE` from `private` to `protected`, changed `numShotsFired` in `Verb_ShootMortartCE` to a `private new int` to work with this change
- Made `CELogger` and `JobGiverUtils_Reload` classes public (So they can be accessed from the separate SOS2Compat Project)

## References

- Closes #3173 

## Reasoning

- Makes CE and SOS2 compatible

## Alternatives

- Leave it unpatched

## Testing

Check tests you have performed:
- [x] Compiles without warnings
- [ ] Game runs without errors
- [ ] (For compatibility patches) ...with and without patched mod loaded
- [ ] Playtested a colony (specify how long)

### CODE REVIEW

**Review by N7Huntsman — CHANGES_REQUESTED** [review]
Torpedo reloading job needs some attention, as it doesn't currently recognize torpedos as usable ammo and reports 'no torpedo' when trying to have a colonist load the launcher. Once loaded however, the torpedos fire correctly.

Shields are correctly intercepting CE projectiles and landed turrets are working.

**Review by N7Huntsman — DISMISSED** [review]
Torpedos are now working, seems good to go.

**Review by MaxDorob — CHANGES_REQUESTED** [review]

File: Source/SOS2Compat/SOS2Compat/SOS2Compat.cs [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+        public IEnumerable<string> GetCompatList()
+        {
+            yield break;
+        }
+        public void PostLoad(ModContentPack content, ISettingsCE _)
+        {
+            BlockerRegistry.RegisterCheckForCollisionCallback(CheckCollision); // For Shields
+            harmony = new Harmony("CombatExtended.Compatibility.SOS2Compat");
+            harmony.PatchAll(Assembly.GetExecutingAssembly());
+        }
```
[the remark this answers is not in the document]
Comment (Rhinous) [reply Source/SOS2Compat/SOS2Compat/SOS2Compat.cs]: Added

File: Source/SOS2Compat/SOS2Compat/SOS2Compat.cs [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+                SoundDef.Named("EnergyShield_Broken").PlayOneShot(new TargetInfo(shield.parent));
+                // Notifs
+                if (shield.parent.Faction != Faction.OfPlayer)
+                {
+                    Messages.Message(TranslatorFormattedStringExtensions.Translate("SoS.CombatShieldBrokenEnemy"), shield.parent, MessageTypeDefOf.PositiveEvent);
+                }
+                else
+                {
+                    Messages.Message(TranslatorFormattedStringExtensions.Translate("SoS.CombatShieldBroken"), shield.parent, MessageTypeDefOf.NegativeEvent);
+                }
```
[the remark this answers is not in the document]
Comment (Rhinous) [reply Source/SOS2Compat/SOS2Compat/SOS2Compat.cs]: Merged the logic into the one method

File: Source/SOS2Compat/SOS2Compat/SOS2Compat.cs [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+                    // Succesful Intercept
+                    // Replicate non-ce SOS2 shield effects
+                    shield.lastInterceptAngle = projectile.DrawPos.AngleToFlat(shield.parent.Position.ToVector3());
+                    shield.lastIntercepted = Find.TickManager.TicksGame;
+                    if (shield.parent.Spawned)
+                    {
+                        FleckMaker.ThrowMicroSparks(shield.parent.DrawPos, shield.parent.Map);
+                        GenExplosion.DoExplosion(projectile.Position, shield.parent.Map, 3, DefDatabase<DamageDef>.GetNamed("ShieldExplosion"), null, screenShakeFactor: 0);
+                    }
+                    exactPosition = BlockerRegistry.GetExactPosition(projectile.OriginIV3.ToVector3(), exactPosition, new Vector3(shield.parent.Position.x, 0, shield.parent.Position.z), shield.radius * shield.radius);
```
[the remark this answers is not in the document]
Comment (Rhinous) [reply Source/SOS2Compat/SOS2Compat/SOS2Compat.cs]: using the result of the intersectionpoint check from further up now

File: Source/SOS2Compat/SOS2Compat/SOS2Compat.cs [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+                return false;
+            }
+
+            shieldPos.y = projectilePos.y; // We just want to check x,z distance
+            // Check if shot is inside shield radius
+            if (Vector3.Distance(shieldPos, projectilePos) > shield.radius)
+            {
+                // outside shield radius so ignore
+                return false;
+            }
```
[the remark this answers is not in the document]
Comment (Rhinous) [reply Source/SOS2Compat/SOS2Compat/SOS2Compat.cs]: Switched to using IntersectionPoint

File: Source/SOS2Compat/SOS2Compat/Building_ShipTurretCE.cs [hunk 5]
```diff
@@ -0,0 +1,1748 @@
...
+using SaveOurShip2;
+using UnityEngine;
+using Verse;
+using Verse.AI;
+using Verse.Noise;
+using Verse.Sound;
+
+namespace CombatExtended.Compatibility.SOS2Compat
+{
+    public class Building_ShipTurretCE : Building_Turret
```
[the remark this answers is not in the document]
Comment (Rhinous) [reply Source/SOS2Compat/SOS2Compat/Building_ShipTurretCE.cs]: Fixed the inheritance

File: Source/SOS2Compat/SOS2Compat/Building_ShipTurretCE.cs [hunk 6]
```diff
@@ -0,0 +1,1748 @@
...
+            else
+            {
+                shipTarget = LocalTargetInfo.Invalid;
+                burstWarmupTicksLeft = 0;
+                if ((mapComp.ShipMapState == ShipMapState.inCombat || GroundDefenseMode) && burstCooldownTicksLeft <= 0)
+                {
+                    TryStartShootSomething(false);
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (Rhinous) [reply Source/SOS2Compat/SOS2Compat/Building_ShipTurretCE.cs]: Fixed the inheritance

**Comment by Rhinous** [comment]
Inheritance should be sorted now to not have a ton of duplicated code, shield logic also updated.

**Review by N7Huntsman — APPROVED** [review]

**Review by N7Huntsman — APPROVED** [review]
Works in testing.

**Comment by Rhinous** [comment]
Split some functionality into #3278 and #3279 will need to add the ground defense mode logic to another PR afterwards

**Comment by Rhinous** [comment]
Cleaned up and added GroundDefenseMode functionality to #3282 making this now obsolete

**Pull request closed without merging**